kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java

```diff
@@ -29,4 +54,18 @@ public static String getStoreTypeOrPlatformDefault(String storeType) {
     public boolean definesKey() {
         return key != null;
     }
+
+    @Override
+    public TlsClientAuth clientAuth() {
+        // We can't just default this in the canonical constructor as it then becomes a nonsensical property of Upstream TLS config
+        return clientAuth == null ? TlsClientAuth.NONE : clientAuth;
+    }
+
+    @VisibleForTesting
+    boolean validateClientAuth() {
+        if (clientAuth() != TlsClientAuth.NONE && Objects.isNull(trust)) {
+            throw new IllegalStateException("ClientAuth enabled but no TrustStore provided to validate certificates");
```

Comment:
Is this condition actually correct?  Why insist that the server overrides his trust anchors to use this feature?  What if he wants to use platform trust?

Using TlsClientAuth.REQUIRED or REQUESTED instructs the TLS server to include a Certificate Request in the TLS handshake.  The client then responds with something that proves he is in possession of the his key (using a signature) and sends his cert.  The server will then verify signature and the check cert against the trust anchor(s).

Reply:
This probably goes back to when I'd made the assumption that providing the trust meant it needed ClientAuth and there was a check on whether trust was provided before setting ClientAuth. Now we're setting trust regardless (even just to PlatformTrustProvider as default) it probably negates this check. 

Your public-CA scenario sounds like a use case for PlatformTrustProvider and ClientAuth REQUIRED.

Reply:
I'd like to hear @SamBarker thoughts.. It has been a long week and my brain is fried!

Reply:
Arguably the check is moot/in appropriate/difficult to satisfy. I hesitated but I was persuaded by @tombentley that when it comes to security its better to err towards manual configuration. Rather than users getting certs validated by the platform trust store when they intend them to be validated by the custom trust store .e.g the proxy is misconfigured but still works by chance.

I do wonder how hard it would be in practice to specify the correct path for the platform trust store.

Reply:
> I hesitated but I was persuaded by @tombentley that when it comes to security its better to err towards manual configuration.

I can see the logic in that idea.

However, if we go down that route, doesn't that mean that `clientAuth` ought to be a property of the `TrustProvider` rather than `Tls`?    That way the configuration model enforces the business rule.

Reply:
@alanrobinson-dwp @SamBarker Following that idea,  I tried pushing clientAuth to the TrustProvider. WDYT?

https://github.com/alanrobinson-dwp/kroxylicious/pull/5

Reply:
I'd been flip flopping for a bit on where to put the clientAuth value and we had a few discussions here. Originally TrustProvider seemed most obvious being that it is only needed if set. We ended up following the broker though which has settings

ssl.client.auth
ssl.keystore.*
ssl.truststore.*

properties, so assuming Tls.java was to replicate the ssl level we landed there. It makes just as much sense in TrustProvider mind.

I had for a split second considered the benefits of having separate Upstream and Downstream TLS classes, but then quickly backed away haha

Reply:
>I'd been flip flopping for a bit on where to put the clientAuth value and we had a few discussions here. Originally TrustProvider seemed most obvious being that it is only needed if set. We ended up following the broker though which has settings

Good to hear its not just me thats having trouble with this, and I really appreciate you taking the time to think it through rather than just pushing up the first thing you came up with.

>properties, so assuming Tls.java was to replicate the ssl level we landed there.

I certainly agree with starting from how does Kafka itself model this as thats what a lot of users will expect but I'm often left a bit confused and frustrated by how Kafka models a lot of config so its not the be all and end all.

>I had for a split second considered the benefits of having separate Upstream and Downstream TLS classes, but then quickly backed away haha

Yeah, I can see why you did but it does echo my thinking from https://github.com/alanrobinson-dwp/kroxylicious/pull/5#discussion_r1868671250 the appeal of Upstream and Downstream is that they are what we want to talk about so in Domain Driven Design terms they are part of our ubiquitous language. Which makes me think its the right direction to be moving. I don't have strong opinions if that happens in this PR or not but we if we want to get to `*stream` types sooner is better so that we have the right shape in the API we release.